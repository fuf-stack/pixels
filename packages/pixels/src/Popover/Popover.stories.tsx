import type { Meta, StoryObj } from '@storybook/react';

import Popover from './Popover';

const meta: Meta<typeof Popover> = {
  title: 'pixels/Popover',
  component: Popover,
};

export default meta;
type Story = StoryObj<typeof Popover>;

export const Default: Story = {
  args: {
    children: 'i am the trigger',
    content: 'i am the content',
  },
};

export const TooMuchContent: Story = {
  args: {
    children: 'i am the trigger',
    placement: 'right',
    content: (
      <>
        {/* eslint-disable-next-line prefer-spread */}
        {Array.apply(null, Array(100)).map((_, i) => (
          // eslint-disable-next-line react/no-array-index-key
          <div key={i}>
            too much content
            <br />
          </div>
        ))}
      </>
    ),
    testId: 'testId42',
  },
};

export const DoubleScrollbar: Story = {
  args: {
    children: 'i am the trigger',
    placement: 'right',
    content: (
      <>
        {/* eslint-disable-next-line prefer-spread */}
        {Array.apply(null, Array(100)).map((_, i) => (
          // eslint-disable-next-line react/no-array-index-key
          <div key={i}>too much content</div>
        ))}
      </>
    ),
    footer: 'i am the footer',
    testId: 'testId42',
    title: 'i am the title',
  },
  render: (args) => (
    <div className="h-[2000px]">
      <Popover {...args} />
    </div>
  ),
};

export const BlockOuterScrollbar: Story = {
  args: {
    children: 'i am the trigger',
    content: (
      <>
        {/* eslint-disable-next-line prefer-spread */}
        {Array.apply(null, Array(100)).map((_, i) => (
          // eslint-disable-next-line react/no-array-index-key
          <div key={i}>too much content</div>
        ))}
      </>
    ),
    footer: 'i am the footer',
    placement: 'right',
    shouldBlockScroll: true,
    testId: 'testId42',
    title: 'i am the title',
  },
  render: (args) => (
    <div className="h-[2000px]">
      <Popover {...args} />
    </div>
  ),
};

export const Title: Story = {
  args: {
    title: 'i am the title',
    children: 'i am the trigger',
    content: 'i am the content',
  },
};

export const Footer: Story = {
  args: {
    footer: 'i am the footer',
    children: 'i am the trigger',
    content: 'i am the content',
  },
};

export const WithCustomTriggerButton: Story = {
  args: {
    content: 'i am the content',
    triggerButtonProps: {
      color: 'success',
      variant: 'shadow',
    },
    children: 'i am a custom trigger',
  },
};

export const AllSections: Story = {
  args: {
    title: 'i am the title',
    footer: 'i am the footer',
    children: 'i am the trigger',
    content: 'i am the content',
  },
};
