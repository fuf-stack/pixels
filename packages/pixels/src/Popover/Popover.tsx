import type { TVClassName, TVProps } from '@fuf-stack/pixel-utils';
import type { PopoverProps as HeroPopoverProps } from '@heroui/popover';
import type { ReactNode } from 'react';
import type { ButtonProps } from '../Button';

import { Button } from '@heroui/button';
import { Popover, PopoverContent, PopoverTrigger } from '@heroui/popover';

import { tv, variantsToClassNames } from '@fuf-stack/pixel-utils';

import ScrollShadow from '../ScrollShadow/ScrollShadow';

// popover styling variants
export const popoverVariants = tv({
  slots: {
    body: 'w-full px-4 py-2',
    content: 'flex max-h-[80vh] flex-col p-0',
    divider: 'm-0 w-full p-0',
    footer: 'w-full px-4 py-2',
    footerWrapper: 'w-full',
    header: 'w-full px-4 pb-1 pt-2 font-semibold',
    headerWrapper: 'w-full',
    trigger: '',
  },
});

type VariantProps = TVProps<typeof popoverVariants>;
type ClassName = TVClassName<typeof popoverVariants>;

export interface PopoverProps extends VariantProps {
  /** child components */
  children?: ReactNode;
  /** CSS class name */
  className?: ClassName;
  /** content of the popover */
  content: ReactNode;
  /** HTML data-testid attribute used in e2e tests */
  contentTestId?: string;
  /** popover footer */
  footer?: ReactNode;
  /** Popover title */
  header?: ReactNode;
  /** use as controlled component  */
  openControlled?: { open: boolean; setOpen: (open: boolean) => void };
  /** placement of the popover relative to its trigger reference */
  placement?: HeroPopoverProps['placement'];
  /** The container element in which the overlay portal will be placed. */
  portalContainer?: HeroPopoverProps['portalContainer'];
  /** Whether to block scrolling outside the popover */
  shouldBlockScroll?: boolean;
  /** HTML data-testid attribute used in e2e tests */
  testId?: string;
  /** When defined a Button will be rendered as trigger (with provided props) instead of unstyled html button */
  triggerButtonProps?: Pick<
    ButtonProps,
    | 'ariaLabel'
    | 'className'
    | 'color'
    | 'disableAnimation'
    | 'disabled'
    | 'size'
    | 'testId'
    | 'variant'
  >;
}

/**
 * Popover component based on [HeroUI Card](https://www.heroui.com//docs/components/popover)
 */
export default ({
  children = null,
  className: _className = undefined,
  content,
  contentTestId = undefined,
  footer = undefined,
  header = undefined,
  openControlled = undefined,
  placement = 'top',
  portalContainer = undefined,
  shouldBlockScroll = undefined,
  testId = undefined,
  triggerButtonProps = undefined,
}: PopoverProps) => {
  // className from slots
  const variants = popoverVariants();
  const className = variantsToClassNames(variants, _className, 'trigger');

  return (
    <Popover
      classNames={className}
      placement={placement}
      portalContainer={portalContainer}
      radius="sm"
      shouldBlockScroll={shouldBlockScroll}
      showArrow
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...(openControlled
        ? { isOpen: openControlled.open, onOpenChange: openControlled.setOpen }
        : {})}
    >
      <PopoverTrigger data-testid={testId}>
        {/* NOTE: type and aria properties are injected by PopoverTrigger */}
        {triggerButtonProps ? (
          // TODO: currently we have to use @heroui/button because
          // passing ref does not work (even with forwardRef in Button)
          // eslint-disable-next-line react/jsx-props-no-spreading
          <Button className={className.trigger} {...triggerButtonProps}>
            {children}
          </Button>
        ) : (
          //  eslint-disable-next-line react/button-has-type
          <button className={className.trigger}>{children}</button>
        )}
      </PopoverTrigger>
      <PopoverContent data-testid={contentTestId}>
        {header && (
          <div className={className.headerWrapper}>
            <div className={className.header}>{header}</div>
            <hr className={className.divider} />
          </div>
        )}
        <ScrollShadow className={className.body}>{content}</ScrollShadow>
        {footer && (
          <div className={className.footerWrapper}>
            <hr className={className.divider} />
            <div className={className.footer}>{footer}</div>
          </div>
        )}
      </PopoverContent>
    </Popover>
  );
};
