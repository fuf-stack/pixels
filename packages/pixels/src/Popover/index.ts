import type { PopoverProps } from './Popover';

import Popover from './Popover';

export type { PopoverProps };
export default Popover;
