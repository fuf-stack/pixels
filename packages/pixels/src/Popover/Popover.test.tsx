import { describe, expect, test } from 'vitest';

import storySnapshots from '@repo/storybook-config/story-snapshots';
import { render } from '@testing-library/react';

import Popover from './Popover';
import * as stories from './Popover.stories';

describe('Story Snapshots', () => {
  storySnapshots(stories);
});

describe('Coverage', () => {
  test('renders no trigger with no children', () => {
    const { container } = render(<Popover content="i am the content" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('controlled popover', () => {
    const { getByText } = render(
      <Popover
        content="i am the content"
        openControlled={{ open: true, setOpen: () => {} }}
      >
        i am the trigger
      </Popover>,
    );
    const trigger = getByText('i am the trigger');
    expect(trigger).toBeInTheDocument();
  });
});
