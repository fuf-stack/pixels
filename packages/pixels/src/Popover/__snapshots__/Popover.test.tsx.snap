// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Coverage > renders no trigger with no children 1`] = `
<button
  aria-expanded="false"
  aria-haspopup="dialog"
  class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
  data-react-aria-pressable="true"
  data-slot="trigger"
  tabindex="0"
  type="button"
/>
`;

exports[`Story Snapshots > AllSections 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="dialog"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    tabindex="0"
    type="button"
  >
    i am the trigger
  </button>
</div>
`;

exports[`Story Snapshots > BlockOuterScrollbar 1`] = `
<div>
  <div
    class="h-[2000px]"
  >
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      data-testid="testId42"
      tabindex="0"
      type="button"
    >
      i am the trigger
    </button>
  </div>
</div>
`;

exports[`Story Snapshots > Default 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="dialog"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    tabindex="0"
    type="button"
  >
    i am the trigger
  </button>
</div>
`;

exports[`Story Snapshots > DoubleScrollbar 1`] = `
<div>
  <div
    class="h-[2000px]"
  >
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      data-testid="testId42"
      tabindex="0"
      type="button"
    >
      i am the trigger
    </button>
  </div>
</div>
`;

exports[`Story Snapshots > Footer 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="dialog"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    tabindex="0"
    type="button"
  >
    i am the trigger
  </button>
</div>
`;

exports[`Story Snapshots > Title 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="dialog"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    tabindex="0"
    type="button"
  >
    i am the trigger
  </button>
</div>
`;

exports[`Story Snapshots > TooMuchContent 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="dialog"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    data-testid="testId42"
    tabindex="0"
    type="button"
  >
    i am the trigger
  </button>
</div>
`;

exports[`Story Snapshots > WithCustomTriggerButton 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="dialog"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none shadow-lg shadow-success/40 bg-success text-success-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    tabindex="0"
    type="button"
  >
    i am a custom trigger
  </button>
</div>
`;
